Adding the diffusion wave routing method to MOSART

liho745 opened the pull request:
This diffusion wave routing method will capture the backwater effects, hence allow better
simulation of river-ocean interactions, particularly the propagation of tidal waves from the ocean
into the rivers. 

[BFB]

rljacob commented:
@hydrotian please review.

hydrotian commented on components/mosart/src/riverroute/RunoffMod.F90:
```diff
@@ -251,6 +255,8 @@ module RunoffMod
      integer , pointer :: iUp(:,:)     ! IDs of upstream units, corresponding to the subbasin ID in the input table
```
A few variables such as `iUp` `nUp` are now defined in `rtmCTL`. Consider removing them from `TUnit`.

liho745 replied:
Done.

liho745 commented on components/mosart/src/riverroute/MOSART_physics_mod.F90:
```diff
@@ -316,22 +372,41 @@ subroutine Euler
+                 numSubSteps = 1 
+                 localDeltaT = Tctl%DeltaT/Tctl%DLevelH2R/numSubSteps
+                 TRunoff%rslp_energy(iunit) = CRRSLP(iunit)
+                 do k=1,numSubSteps
```
I'd prefer to keep it there, in case there is a future need for improving the numerical scheme.

liho745 commented on components/mosart/src/riverroute/MOSART_physics_mod.F90:
```diff
@@ -316,22 +372,41 @@ subroutine Euler
-                temp_erout = temp_erout + TRunoff%erout(iunit,nt) ! erout here might be inflow to some downstream subbasin, so treat it differently than erlateral                
-             end do
-             temp_erout = temp_erout / TUnit%numDT_r(iunit)
+             if(Tctl%RoutingMethod==1) then  ! local stepping method only applicable for kinamatic wave routing method
```
Done. I've reduced the total number of routing options to 2 now. 1--KW and 2--DW. Please be aware that I also changed the relevant inuncation code for consistency.

liho745 commented:
@hydrotian I've addressed your comments and pushed my changes to the E3SM repository. I've done some testing at my end by running a global tri-grid case for about 2.5 calendar years. The spatial pattern of simulated streamflow seems reasonable, and the global water balance is preserved all the time. Please review again.

bishtgautam commented:
@liho745 This PR broke most of the tests (see https://my.cdash.org/viewTest.php?onlyfailed&buildid=1987652). Did you run tests before issuing the PR? 
Please create baselines of the test and perform comparisons against those baselines. Instructions on how to do this are at https://gist.github.com/bishtgautam/31a73b933e2492698bccfc7b55acf01e

liho745 commented:
@bishtgautam Thanks for the update. I will work on it with @hydrotian's help.

liho745 commented:
@donghuix I've made some changes to the sub-cycling. @hydrotian did the tests and all passed. Are you able to perform another round of scientific testing/validation to make sure the science is still good? Please note that, now you will only need to set the routingmethod = 2 (in user_nl_mosart), and no need to set DLEVELH2R anymore.

donghuix commented:
@liho745 Sure, I can validate it. Is it already in the master now? I saw the "Merging is blocked", should I use your branch hongyili/mosart/mosart-dw?

liho745 commented:
@donghuix Thanks! Yes, you should use my branch instead.

liho745 commented:
@donghuix The changes I made also slightly affect the kinematic wave method. Could you test/validate the kinematic wave method as well?

donghuix commented:
@liho745 Ok, I will run for both methods to see the differences.

donghuix commented:
@liho745 I validated your branch with Susquehanna basin and force MOSART with Ming Pan's runoff data. The attached time series comparison shows good comparison with observation at monthly scale. The spatial plot is the averaged streamflow for 30 years (1979-2008), and the differences between DW and KW are very small. 

<img width="1504" alt="Screen Shot 2021-04-08 at 10 37 56 PM" src="https://user-images.githubusercontent.com/64861893/114133717-55292c00-98bb-11eb-9de7-45cfeb199d1a.png">
<img width="1462" alt="Screen Shot 2021-04-08 at 10 38 01 PM" src="https://user-images.githubusercontent.com/64861893/114133720-565a5900-98bb-11eb-9fb9-7e7153b158dd.png">

liho745 commented:
@donghuix Thanks! The results look indeed quite good. It is not surprising that the difference at the monthly scale is small, and I'd expect a more noticeable difference at the hourly/daily scale, or after we incorporate the tidal effects. 

@hydrotian Would you expect more scientific validation or code review at this point? If not, given that all tests passed at your end, perhaps we can suggest the merging move forward?

hydrotian commented:
@liho745 @bishtgautam Yes I think we don't need further tests and we should try merging it to the master. I do see there's a ci/circleci build error. Not sure if that will affect the merging.

bishtgautam commented:
Is the comparison of all regression tests with their baselines BFB?

hydrotian commented:
Yes they are BFB.
